### ISSUE TITLE
NLnet grant application for federation (deadline October 1st, 2021)

### ISSUE THREAD

**Issue by techknowlogick** [go-gitea/gitea#16827]
Bonjour,

During the [videoconference concluding the generic grant application](https://github.com/go-gitea/gitea/issues/16518), it was decided that @zeripath and @techknowlogick will apply to the [NLnet Next Generation Search and Discovery](https://nlnet.nl/discovery/) call for proposal for a total amount of 50K€. Should the proposal be accepted the work could begin around January 1st, 2022.

This issue will contain the draft of the proposal, its final version and all communications with NLnet (only redacted for privacy) for the sake of transparency.

The action items are distributed as follows:

* @dachary :
  * :white_check_mark: Write down a [NGI Zero Discovery NLnet](https://nlnet.nl/discovery/) draft [application](https://demo.hedgedoc.org/gIyguVYPQyqQnnsw39sEhQ#), reusing material from the generic grant application
  * Select tasks that can fit in the NLnet budget
* @techknowlogick: provide a bio in [the generic grant application](https://demo.hedgedoc.org/K8IDhSxJShOCOWEKILGZeA?view)
* @zeripath @techknowlogick: provide an hourly/daily rate
* @zeripath @techknowlogick: proofread the [NLnet application](https://demo.hedgedoc.org/gIyguVYPQyqQnnsw39sEhQ#) draft

The proposal will be submitted by @zeripath as the primary contact.

Cheers

---
Posted on behalf of [@dachary](https://forum.fedeproxy.eu/u/dachary), [source](https://forum.fedeproxy.eu/t/nlnet-grant-application-for-federation-in-gitea-deadline-october-1st-2021/354)

[NLnet.md](https://github.com/go-gitea/gitea/files/7060570/Untitled.md)

**Comment by pilou-** [comment]
Posted on [behalf](https://forum.fedeproxy.eu/t/nlnet-grant-application-for-federation-in-gitea-deadline-october-1st-2021/354/6) of Loic.

---

Bonjour,

This week-end NLnet sent the following message, which motivated me to draft the workplan sooner rather than later:

> **Last call for NGI0 Discovery approaching**
> 
> Please note that we are rapidly approaching the **last call round** from [NGI Zero Discovery](https://nlnet.nl/discovery/). Currently over 90% of the available budget is already allocated to 135 projects, with the August call still being processed. This means that if you are nearing the end of your existing grant and were considering applying for a follow up project - you probably do not want to miss the October 1st deadline...


Here is [a tentative workplan](https://demo.hedgedoc.org/_1U27XNKQ3-yK-Sm_-gcYA?view)  for a total budget of 50,000€ with an hourly rate of £40 / 45€ / 70$ Canadian with 7 hours per day i.e. 315€ per day that is approximately 160 days of work or 8 month (20 work days per month).  @zeripath @techknowlogick this is merely a guide to get you started: since you're going to be the one doing the work it must ultimately be something you feel you can easily accomplish in the timeframe. And the hourly rate must also be adjusted to match your needs. A few notes:

* This is **not a full time commitment, it can stretch over a longer period of time depending on your availability**. The payments of the grant are not time based, they happen when a public proof of the deliverable completion is available.
* While the workplan is based on the one included in the [generic grant application](https://demo.hedgedoc.org/K8IDhSxJShOCOWEKILGZeA?view), I stripped it down to the very minimum (as far as I can do that given my limited knowledge). I tried my best to focus on what would be a good foundation for gitea to build on. In other words I tried to **keep the bigger items, those that really require funding to implement** because they need days if not weeks to land.
* I suggest you rework this workplan so that you are **100% convinced it will be very easy to finish**. In my experience this is the key to success.
* The **Web Client and User Research task lists are not included because they are unlikely to fit in the budget**. The grant will be enough to make significant progress with the infrastructure changes required for federation. But committing to deliver user facing changes in that timeframe is probably too ambitious.

What do you think?

[plan.md](https://forum.fedeproxy.eu/uploads/short-url/qnuJViB066gm6Gziz3DDymO3fNW.md) (6.0 KB)

**Comment by pilou-** [comment]
Posted on [behalf](https://forum.fedeproxy.eu/t/nlnet-grant-application-for-federation-in-gitea-deadline-october-1st-2021/354/15) of Loic.

---

@techknowlogick @zeripath a gentle reminder: you have pending action items :-) The deadline is in **15 days** and it would be wise to finish it as soon as possible to avoid any last minute troubles.

* Fill your bio & motivation (@techknowlogick) in [the grant application](https://demo.hedgedoc.org/gIyguVYPQyqQnnsw39sEhQ#)
* Decide on your hourly/daily rate (in this thread or in [the grant application](https://demo.hedgedoc.org/gIyguVYPQyqQnnsw39sEhQ#))
* Review and update [the draft workplan](https://demo.hedgedoc.org/_1U27XNKQ3-yK-Sm_-gcYA?view#)

It should not take more than one hour of your time.

**Comment by pilou-** [comment]
Posted on behalf of [Loic](https://forum.fedeproxy.eu/t/nlnet-grant-application-for-federation-in-gitea-deadline-october-1st-2021/354/16)

---

I have good news :slight_smile:  It turns out the workplan is **not binding**. What it means is that this workplan is just an indication that there exists a workplan already, instead of not. Assuming the proposal is accepted, the actual workplan as well as the cost of each subtask will be negotiated roughly one month **after** the acceptance. It makes sense: plans and personal situations can change between the date of the submission and the date when the work begins, three months later.

Bottom line,  @zeripath has no work item left :tada: And there is just one item left for @techknowlogick : filling your bio & motivation in [the grant application](https://demo.hedgedoc.org/gIyguVYPQyqQnnsw39sEhQ#).

**Comment by pilou-** [comment]
Posted on behalf of [Loic](https://forum.fedeproxy.eu/t/nlnet-grant-application-for-federation-in-gitea-deadline-october-1st-2021/354/21)

---

@techknowlogick I copy/pasted the bio you sent via email. The application is complete :tada: Since you agreed with @zeripath on having the same daily rate, I assumed you would be comfortable with an hourly rate of 60€ and wrote down a daily rate of 480€ per day in the proposal. Feel free to change that if this is not correct.

I wrote a letter of recommendation that essentially says I support this grant application because it will be an important step forward to make fedeproxy redundant (being redundant is the ultimate goal of fedeproxy, the measure of success).

[2021-10-gitea-nlnet-loic-dachary-recommendation.txt](https://forum.fedeproxy.eu/uploads/short-url/294ncn5wp8tRY0XAsaoBDdP2tik.txt) (1.5 KB)

Unless you have an objection, I will submit the proposal at https://nlnet.nl/thema/NGIZeroDiscovery.html on your behalf **September 27th, 2021**, together with the letters of recommendation. Since @zeripath is the primary contact, he will receive the confirmation and from that point on it will be in your hands, I will no longer be involved.

[2021-10-gitea-nlnet-application.txt](https://forum.fedeproxy.eu/uploads/short-url/etQeCCO9FpfvKe1gkDLyiimrb2p.txt) (8.5 KB)
[2021-10-gitea-nlnet-workplan.txt](https://forum.fedeproxy.eu/uploads/short-url/qnuJViB066gm6Gziz3DDymO3fNW.md) (6.0 KB)

**Comment by rektide** [comment]
This is a core capability that will let developers find new ways to work together & to share work. It's also a core step to making the act of coding more visible, & more tangible. I could not be more excited for this work. Good luck on this grant application.

**Comment by pilou-** [comment]
Posted on behalf of [Loic](https://forum.fedeproxy.eu/t/nlnet-grant-application-for-federation-in-gitea-deadline-october-1st-2021/354/25).
<hr>

There was no objection, I'll post the proposal today at [NLnet; NGI Zero Discovery](https://nlnet.nl/thema/NGIZeroDiscovery.html) on behalf of @zeripath. I made minor wording changes, the final version is attached below for archive.

[2021-10-gitea-nlnet-application.txt|attachment](https://forum.fedeproxy.eu/uploads/short-url/jWirMy7Mqr10Ff1Ce3hZ4h6EOdI.txt) (8.5 KB)
[2021-10-gitea-nlnet-workplan.txt|attachment](https://forum.fedeproxy.eu/uploads/short-url/qnuJViB066gm6Gziz3DDymO3fNW.md) (6.0 KB)
[2021-10-gitea-nlnet-loic-dachary-recommendation.txt|attachment](https://forum.fedeproxy.eu/uploads/short-url/294ncn5wp8tRY0XAsaoBDdP2tik.txt) (1.5 KB)

From that point on **I will not be involved anymore**. I'm very happy about the work that was done. Even if this grant application is denied, there is excellent material to [answer future calls](https://github.com/go-gitea/gitea/issues/16518) and I'm convinced funding will be granted eventually. It was made in a 100% transparent way, which is very rare and valuable to the Free Software community. I already stated that [it was a precondition for my participation](https://forum.fedeproxy.eu/t/grant-application-for-federation-in-gitea/284) and I'd like to express my gratitude for everyone involved: it is an unusual requirement and it went very smoothly.

**Comment by zeripath** [comment]
Thank you Loïc for all of your work on this!

**Comment by ShadowJonathan** [comment]
I heard from someone that this development, even though the grant wasn't given, is still being continued, where can i find information on that?

**Comment by melvincarvalho** [comment]
I think this would be relatively easy to get going

1. Have a JSON(-LD) blob in your gitea profile page, embedded in a <SCRIPT> tag

2. This blob contains your profile information, your nick (`@user@gitea.example`), and, importantly a nested link to your SSH keys.  An important design consideration in the JSON blob is to allow different objects e.g. (`#me` - Person, `#key` - public key, `""` - document meta data)

3. Start to allow federated features such as single-sign-on, pull requests, where your ssh key can be used to verify your federated identity

After this proof of concept is working, expand out the federation features

I'll add that since yesterday, you can now sign commits with your SSH key, that's already a big step forward, and this style can be expanded to create global social coding

**Comment by aschrijver** [comment]
Cross-ref to prior ActivityPub design discussion: https://github.com/go-gitea/gitea/issues/14186

**Comment by melvincarvalho** [comment]
@aschrijver thank you for the pointer!

Let me add this tag for those discovering this thread:

https://github.com/go-gitea/gitea/labels/theme%2Ffederation

From my experience, doing the profile pages well gives the best pay-off, in terms of federation.  That's because it acts as a means of orchestration for everything else.  They are also really hard to future-proof, so the first version of the design is likely to stick around for a long time.  The simple step of adding your public key to your profile is a major improvement (or standarizing on the `.keys` pattern which can be tagged on to the username).  Will continue design discussion in the relevant threads

**Comment by madnificent** [comment]
Thank you for the work on this!  Is this the issue we should track for development on this, or is there another tracking issue I should follow?

**Comment by techknowlogick** [comment]
Apologies for keeping mum on this, however we can now say we have been accepted into NLnet. https://nlnet.nl/project/Gitea/

**Comment by melvincarvalho** [comment]
> Apologies for keeping mum on this, however we can now say we have been accepted into NLnet. https://nlnet.nl/project/Gitea/

Congrats!

I can commit to running a testing server.  I will try and help overall too with milestones, design and code.  I have a use case in mind which is to do with signing git commits.

**Comment by aschrijver** [comment]
FYI There's great interest. The announcement is also [discussed on Hacker News](https://news.ycombinator.com/item?id=29830061)
